airbyte_cdk/sources/declarative/decoders/composite_decoder.py

```diff
@@ -0,0 +1,127 @@
+import gzip
+import json
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Any, Generator, Iterable, MutableMapping, Optional, Protocol
+
+import pandas as pd
+import requests
+from numpy import nan
+
+from airbyte_cdk.sources.declarative.decoders.decoder import Decoder
+
+
+class BufferedDataInput(Protocol):
+    """
+    at least read to _n bytes method should be supported to align with BufferedReader and avoid OOM.
+    """
+
+    def read(self, __n: int) -> bytes: ...
+
+
+@dataclass
+class Parser(ABC):
+    inner_parser: Optional["Parser"] = None
+
+    @abstractmethod
+    def parse(
+        self, data: BufferedDataInput | Iterable, *args, **kwargs
```

Comment:
I'm a bit confused by this interface. Does that mean that all the users with have to check the `data` parameter as such:
```
    def parse(
        self, data: BufferedDataInput | Iterable, *args, **kwargs
    ) -> Generator[MutableMapping[str, Any], None, None]:
        if isinstance(data, BufferedDataInput):
            < do X >
        elif isinstance(data, Iterable):
            < do Y >
        else:
            raise ValueError(f"Unexpected type {type(data)}")
```

I also don't see the usage of `BufferedDataInput` in the sample parsers you provided below, maybe I missed it. Can you show me how this would be used?

Reply:
> Does that mean that all the users with have to check the data parameter as such:

Usually not, cause I expect user to choose _valid_ parser sequence.

> I also don't see the usage of BufferedDataInput in the sample parsers you provided below, maybe I missed it. Can you show me how this would be used?

We start processing from passing `response.raw` in 
```python3
self.parser.parse(data=response.raw)
```
which in fact is readable (buffered) object.

any decompressor, like gzip or zip returns decompressor object , which in turn is also readable (buffered).

Reply:
Ok! I really like the idea of having a buffered reader as a part of the interface. Maybe it should even be a by more meaning that the file-based [CsvParser is relying on more methods than just "read"](https://github.com/airbytehq/airbyte-python-cdk/blob/57e1b524e2c55b2c1e8b80da38c0cfda3c6a385a/airbyte_cdk/sources/file_based/file_types/csv_parser.py#L151). 

So I think my question now is: can we remove `Iterable` from the interface?

Reply:
I guess yes, refactored interface to `BufferedIOBase`